### ISSUE TITLE
Category swatches not working in Magento 2.3

### ISSUE THREAD

**Issue by ladle3000** [magento/magento2#19688]
### Preconditions (*)

1. Magento 2.3
2. 3rd Party Themes & Modules disabled
3. Running in developer mode.

### Steps to reproduce (*)

1. Visit a category page where swatches would properly show on magento 2.2.5 (upgraded directly from this) and ran recreate catalog images requirement that were listed in 2.2.6 release notes

### Expected result (*)

1. Expecting swatches to show properly on catalog pages.

### Actual result (*)

1. Swatches missing on category pages http://nimb.ws/OpLlN1 and have removed pub/static/frontend and flushed caches several times.
2. Error in chrome console:

swatch-renderer.js:555 Uncaught TypeError: Cannot read property 'swatchImage' of undefined
    at Object.<anonymous> (swatch-renderer.js:555)
    at Function.each (jquery.js:370)
    at $.(/shop/anonymous function).(anonymous function)._RenderSwatchOptions (https://devtgn.com/pub/static/version1544449453/frontend/Magento/luma/en_US/Magento_Swatches/js/swatch-renderer.js:496:15)
    at $.(/shop/anonymous function).(anonymous function)._RenderSwatchOptions (https://devtgn.com/pub/static/version1544449453/frontend/Magento/luma/en_US/jquery/jquery-ui.js:402:25)
    at Object.<anonymous> (swatch-renderer.js:395)
    at Function.each (jquery.js:370)
    at $.(/shop/anonymous function).(anonymous function)._RenderControls (https://devtgn.com/pub/static/version1544449453/frontend/Magento/luma/en_US/Magento_Swatches/js/swatch-renderer.js:392:15)
    at $.(/shop/anonymous function).(anonymous function)._RenderControls (https://devtgn.com/pub/static/version1544449453/frontend/Magento/luma/en_US/jquery/jquery-ui.js:402:25)
    at $.(/shop/anonymous function).(anonymous function)._init (https://devtgn.com/pub/static/version1544449453/frontend/Magento/luma/en_US/Magento_Swatches/js/swatch-renderer.js:310:22)
    at $.(/shop/anonymous function).(anonymous function)._init (https://devtgn.com/pub/static/version1544449453/frontend/Magento/luma/en_US/jquery/jquery-ui.js:402:25)

**Comment by magento-engcom-team** [comment]
Hi @snoroozi. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@snoroozi do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by ladle3000** [comment]
Yes, I've provided all info @magento-engcom-team

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @snoroozi thank you for you report, i'm not able to reproduce following steps you described.
As i can see,  you have undefined variable "swatchImage.width" so problem may be related to your some swatch images of product that have no this attribute.

**Comment by ladle3000** [comment]
@engcom-backlog-nazar thanks for your comments. However I see nowhere to set the width as you suggest. Can you point me to it?

As I've mentioned, I've tried bundled themes, and I don't see this setting in their configuration.

**Comment by ladle3000** [comment]
Actually, this is a CORE bug as the setting that triggers this is to use product image as swatch if possible.  Confirmed by Mirasvit as well:

"Locally we have switched the attribute option "Use Product Image for Swatch if Possible" to "Yes" and we get the same error in the browser console, saying "Uncaught TypeError: Cannot read property 'swatchImage' of undefined". This error happens only on category pages because Magento initializes the swatches without required option "jsonSwatchImageSizeConfig".

To fix it you need to add the string ""jsonSwatchImageSizeConfig": <?php /* @noEscape */ echo $block->getJsonSwatchSizeConfig() ?>" to the file "vendor/magento/module-swatches/view/frontend/templates/product/listing/renderer.phtml""

**Comment by joebordo** [comment]
I have the Same problem. Thanks Snoroozi for the fix, i will try now!

**Comment by ladle3000** [comment]
@joebordo I even opened another ticket with the added prerequisites and they say it doesn't happen on clean.  What all extensions do you use?

**Comment by joebordo** [comment]
@snoroozi 

EDIT: I use it on Fresh install on 2.3.
Your fix worked but u should edit your post for the correct renderer file. It should be folder listing not view. I added this string after Mediacallback and now it seems to work.

`"jsonSwatchImageSizeConfig": <?php /* @noEscape */ echo $block->getJsonSwatchSizeConfig() ?>`

File to edit:
`vendor/magento/module-swatches/view/frontend/templates/product/listing/renderer.phtml"`

Thanks btw! for some reason i tought i was doing  something wrong wasted hours...before i found your post

**Comment by joebordo** [comment]
this is how my renderer looks like. I see the string is already there. 

`<?php
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */
// @codingStandardsIgnoreFile
//"swatchRenderer": {
?>
<?php /** @var $block \Magento\Swatches\Block\Product\Renderer\Configurable */ ?>
<div class="swatch-opt" data-role="swatch-options"></div>

<script type="text/x-magento-init">
    {
        "[data-role=swatch-options]": {
            "Magento_Swatches/js/swatch-renderer": {
                "jsonConfig": <?= /* @escapeNotVerified */ $swatchOptions = $block->getJsonConfig() ?>,
                "jsonSwatchConfig": <?php /* @escapeNotVerified */
                    echo $swatchOptions = $block->getJsonSwatchConfig(); ?>,
                "mediaCallback": "<?= /* @escapeNotVerified */ $block->getMediaCallback() ?>",
                "gallerySwitchStrategy": "<?php /* @escapeNotVerified */ echo $block->getVar('gallery_switch_strategy',
                    'Magento_ConfigurableProduct') ?: 'replace'; ?>",
                "jsonSwatchImageSizeConfig": <?php /* @noEscape */ echo $block->getJsonSwatchSizeConfig() ?>
                
            }
        },
        "*" : {
            "Magento_Swatches/js/catalog-add-to-cart": {}
        }
    }
</script>
`

**Comment by ladle3000** [comment]
@joebordo updated. We need to get this acknowledged by the team. Do you have any further info from your side of what particular setup causes this?

I opened this one too https://github.com/magento/magento2/issues/19830 and they still closed it.

**Comment by joebordo** [comment]
@snoroozi 

I have a clean install of magento 2.3 with no extensions on php7.2 all required extensions. 

I just created a new attribute "Color" Added 3 Color Types". Same thing i did on 2.2.x...without problem.

After adding the above fix, swatches show up and there is no error in console log.

**Comment by hostep** [comment]
Hi guys

I was able to reproduce this on a vanilla 2.3.0 installation after trying to figure out how these swatches work and then it was very easy to reproduce (don't understand how @engcom-backlog-nazar was not able to reproduce, but whatever).
I'm not going to add the exact steps to reproduce here, because the bug is already fixed on the `2.3-develop` branch, so I'm not going to put more effort in this comment then strictly needed :)

Anyway, the fix is here: https://github.com/magento/magento2/commit/c74b0bbf397#diff-f81ffaa60c1bd7810e67f1f63a96e692, I don't know when exactly it will be released, maybe in 2.3.1, but that's unknown to me.

Hope this helps :)

**Comment by ghost** [comment]
@hostep Hi, In our new workflow we try to reproduce first in 2.3-develop branch, and in precondition -> Magento 2.3 not a 2.3.0 :)

**Comment by ladle3000** [comment]
Glad to see some responses here finally @engcom-backlog-nazar - you must have not been able to reproduce on the other ticket because you started with develop branch.

So you going to reopen and merge issues?

**Comment by hostep** [comment]
@snoroozi: the issue is already fixed by the core Magento team and will be rolled out in one of the upcoming releases, so it makes no sense to re-open the issue I think. Thanks for the report though!

**Comment by engcom-backlog-nazar** [comment]
@hostep but who then close this one ? :)

**Comment by ladle3000** [comment]
@hostep I guess I mean moreso to add acknowledgment and reference the fix for those that run across the issue on github.
